flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/BroadcastRecordWriter.java

```diff
@@ -130,7 +132,7 @@ public BufferBuilder requestNewBufferBuilder(int targetChannel) throws IOExcepti
 
 		BufferBuilder builder = super.requestNewBufferBuilder(targetChannel);
 		if (randomTriggered) {
-			addBufferConsumer(builder.createBufferConsumer(), targetChannel);
+			addBufferConsumer(randomConsumer = builder.createBufferConsumer(), targetChannel);
```

Comment:
I think we should not construct the consumer before filling the `BufferBuilder`, otherwise the random emitted latency marker is also visible for all the other channels, which breaks the previous assumption.

Reply:
We have to or else the underlying buffer is not reference counted properly.
The idea is to keep the `randomConsumer` and use `copy` to use the then updated reader index. 
If the random event spans multiple buffers, only the last `randomConsumer` is effectively used.

Reply:
Yeah,  your thought is right. But my concern is that the current changes would make the latency marker visible for all the channels, not only for one random channel if my understanding is right.

I mean the current created `randomConsumer` should have `0` reader position, then all the copies should also have the `0` reader position, which means they can read the latency marker because it is filled in the `BufferBuilder` afterwards.

I guess the proper way might be we only create one consumer for the random channel before filling the latency marker. After that, we create/copy the separate consumers with increased references for all the left channels, then the reader position for these copied  consumers should overtake the length of latency marker?

Reply:
You are right, currently LM gets replicated, which is not good. I was expecting `#copy` to pick up the current builder position for some reason.

Coming back to this bug, the issue is that the underlying memory segment will be recycled as soon as the first buffer consumer (with LM) is being consumed. That causes buffers to be effectively skipped and corrupted streams.

@pnowojski and I was also questioning why `BufferBuilder` only holds the memory segment and not the buffer. By definition calling `#createBufferConsumer` more than once will create the aforementioned race condition.

It would be solved if `BufferBuilder` holds the buffer instead of the raw memory segment as a buffer would only be recycled when the last buffer consumer has been closed. However, it seems so obvious that there must be a reason for the current implementation.

Reply:
Creating `Buffer` in `BufferBuilder` would require two more `volatile` access when bumping the reference count on the per buffer path. I would expect this to be visible in our benchmarks with high number of network channels and frequent flushing (`1000,1ms`). I'm not sure, I might be wrong but I think I was even benchmarking this on Nico's request during the initial development of low latency changes.

Reply:
Okay I added a proposal that strives for a middle ground: buffer is owned by `BufferBuilder` but not initialized before first buffer consumer is being created. I suspect though that now `buffer` needs to be `volatile` :(.

Reply:
Alternatively, we "just" need to add a specific API that copies a `BufferConsumer` but set the positionMarker to that of a given `BufferBuilder`.

```
	public BufferConsumer copyAndSync(BufferBuilder builder) {
		return new BufferConsumer(buffer.retainBuffer(), builder.positionMarker,
			builder.positionMarker.cachedPosition);
	}
```